packages/emnapi/src/simple-async-operation.ts

```diff
@@ -1,119 +1,51 @@
-      if (w) {
-        const env = $makeGetValue('w', 0, '*')
-        $makeSetValue('w', POINTER_SIZE * 4, '0', '*') // tid
-        const complete = $makeGetValue('w', POINTER_SIZE * 2, '*')
-        if (complete) {
-          const envObject = emnapi.envStore.get(env)!
-          const scope = emnapi.openScope(envObject, emnapi.HandleScope)
-          try {
-            envObject.callIntoModule((_envObject) => {
-              const hint = $makeGetValue('w', POINTER_SIZE * 3, '*')
-              emnapiGetDynamicCalls.call_vpip(complete, env, napi_status.napi_ok, hint)
-            })
-          } catch (err) {
-            emnapi.closeScope(envObject, scope)
-            throw err
-          }
-          emnapi.closeScope(envObject, scope)
+      postMessage({
+        emnapiAsyncSend: {
+          callback,
+          data
         }
-      }
-    }
-    if (ENVIRONMENT_IS_NODE) {
-      worker.on('message', worker._emnapiAsyncWorkListener)
+      })
     } else {
-      worker.addEventListener('message', worker._emnapiAsyncWorkListener, false)
+      setTimeout(() => {
+        emnapiGetDynamicCalls.call_vp(callback, data)
+      })
     }
-  }
-}
-
-function napi_cancel_async_work (env: napi_env, work: number): napi_status {
-// #if USE_PTHREADS
-  return emnapi.checkEnv(env, (envObject) => {
-    return emnapi.checkArgs(envObject, [work], () => {
-      $from64('work')
-      const tid = $makeGetValue('work', POINTER_SIZE * 4, '*')
-      const workQueueIndex = emnapiAsyncWorkerQueue.indexOf(work)
-      if (tid !== 0 || workQueueIndex === -1) {
-        return envObject.setLastError(napi_status.napi_generic_failure)
-      }
-
-      emnapiAsyncWorkerQueue.splice(workQueueIndex, 1)
-      const complete = $makeGetValue('work', POINTER_SIZE * 2, '*')
-      if (complete) {
-        setTimeout(() => {
-          const envObject = emnapi.envStore.get(env)!
-          const scope = emnapi.openScope(envObject, emnapi.HandleScope)
-          try {
-            envObject.callIntoModule((_envObject) => {
-              const hint = $makeGetValue('work', POINTER_SIZE * 3, '*')
-              emnapiGetDynamicCalls.call_vpip(complete, env, napi_status.napi_cancelled, hint)
-            })
-          } catch (err) {
-            emnapi.closeScope(envObject, scope)
-            throw err
-          }
-          emnapi.closeScope(envObject, scope)
-        })
-      }
-
-      return envObject.clearLastError()
-    })
-  })
-// #else
-  return _napi_set_last_error(env, napi_status.napi_generic_failure, 0, 0)
-// #endif
-}
-
-emnapiImplement('_emnapi_queue_async_work_js', _emnapi_queue_async_work_js, ['$PThread', '$emnapiAsyncWorkerQueue', '$emnapiGetDynamicCalls'])
-emnapiImplement('napi_cancel_async_work', napi_cancel_async_work, ['$emnapiAsyncWorkerQueue', '$emnapiGetDynamicCalls', 'napi_set_last_error'])
+  },
+  _emnapi_async_send__postset:
+    'PThread.unusedWorkers.forEach(emnapiAddSendListener);' +
```

Comment:
Hm, if you switched to libuv / native threadpool, why do you still need to access Emscripten's internals here? (I can probably figure out, but quicker to ask.)

Reply:
`_emnapi_async_send` emulates `uv_async_send`. Here still need to use JavaScript to send events to the main thread, or perform an asynchronous operation in the event loop. Is there any better choice to do `uv_async_send` with emscripten API on C side?

Reply:
I'm not very familiar with libuv tbh. What does it do here - does it also use custom JS for sending those via V8?

Reply:
> does it also use custom JS for sending those via V8?

It use eventfd(linux) / IOCP(windows) to wake up main thread uv loop to call the complete callback, not use v8.

Currently emnapi's async work implementation is almost a copy of Node.js, but it's difficult for me to port libuv's `async.c` which includes the thread pool async dispatching functions, so I think it's ok to leave this part be JavaScript.

Reply:
Hm, is it always for sending work to the main thread or to any? If just to main, [`MAIN_THREAD_ASYNC_EM_ASM`](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_thread_em_asm#c.MAIN_THREAD_ASYNC_EM_ASM) could be a better option.

Reply:
`MAIN_THREAD_ASYNC_EM_ASM` is not the expect one, it should always non-blocking.

> // If the calling thread is the main browser thread, then the specified JavaScript code is executed
> // synchronously.

`emscripten_dispatch_to_thread_async` internal also use `setTimeout`, right?

https://github.com/emscripten-core/emscripten/blob/main/system/lib/pthread/library_pthread.c#L638

and internally using the `do_dispatch_to_thread` also may call the code synchronously.

https://github.com/emscripten-core/emscripten/blob/main/system/lib/pthread/library_pthread.c#L383

Reply:
> is not the expect one, it should always non-blocking.

Not sure what you mean. I man, it _is_ non-blocking, unless you run it on the same thread, but why would you use it on the same thread anyway?

Reply:
because napi_call_threadsafe_function also can call uv_async_send, in the main thread

Reply:
https://github.com/toyobayashi/emnapi/blob/feat-threadpool/packages/emnapi/src/emnapi.c#L631

https://github.com/toyobayashi/emnapi/blob/feat-threadpool/packages/emnapi/src/emnapi.c#L604

Reply:
So, I mean

```js
function _emnapi_async_send (callback, data) {
  if (ENVIRONMENT_IS_PTHREAD) {
    // not main thread
    postMessage(...)
  } else {
    // main thread should delay by setImmediate or Promise
    setImmediate(() => {
      call_vp(callback, data)
    })
  }
}
```

Reply:
Another thing you can do for now to keep it simple is just 

```js
MAIN_THREAD_ASYNC_EM_ASM({
  setImmediate(() => { ... });
});
```

which is not perfect perf-wise, but it would be simple and will ensure that the function is always invoked asynchronously.

Reply:
Thank you for your patient explanation. I will take a look tomorrow after work, it's midnight in my time zone now

Reply:
No worries at all, thanks!